subsys/bluetooth/host/classic/l2cap_br.c

```diff
@@ -293,1712 +405,5198 @@ static void cancel_data_ready(struct bt_l2cap_br_chan *br_chan)
+	    (br_chan->tx.mode == BT_L2CAP_BR_LINK_MODE_STREAM)) {
+		if (br_chan->tx.extended_control) {
+			outstanding_frames = outstanding_frames % BT_L2CAP_EXT_CONTROL_SEQ_MAX;
+			ack_frames = ack_frames % BT_L2CAP_EXT_CONTROL_SEQ_MAX;
+		} else {
+			outstanding_frames = outstanding_frames % BT_L2CAP_CONTROL_SEQ_MAX;
+			ack_frames = ack_frames % BT_L2CAP_CONTROL_SEQ_MAX;
+		}
+	} else if ((br_chan->tx.mode == BT_L2CAP_BR_LINK_MODE_RET) ||
+		   (br_chan->tx.mode == BT_L2CAP_BR_LINK_MODE_FC)) {
+		outstanding_frames = outstanding_frames % BT_L2CAP_CONTROL_SEQ_MAX;
+		ack_frames = ack_frames % BT_L2CAP_CONTROL_SEQ_MAX;
+	}
 
-	__ASSERT(pdu, "signaled ready but no PDUs in the TX queue");
+	outstanding_count = bt_l2cap_br_get_outstanding_count(br_chan);
+	__ASSERT(outstanding_count == outstanding_frames,
+		 "Mismatch between window and seq %zu - %zu", outstanding_count,
+		 outstanding_frames);
 
-	if (bt_buf_has_view(pdu)) {
-		LOG_ERR("already have view on %p", pdu);
-		return NULL;
+	if (ack_frames <= outstanding_frames) {
+		return true;
 	}
 
-	/* We can't interleave ACL fragments from different channels for the
-	 * same ACL conn -> we have to wait until a full L2 PDU is transferred
-	 * before switching channels.
-	 */
-	bool last_frag = amount >= pdu->len;
+	return false;
+}
 
-	if (last_frag) {
-		LOG_DBG("last frag, removing %p", pdu);
-		__maybe_unused struct net_buf *b = k_fifo_get(&br_chan->_pdu_tx_queue, K_NO_WAIT);
+static void l2cap_br_sdu_is_done(struct bt_l2cap_br_chan *br_chan, struct net_buf *sdu, int err)
+{
+	bt_conn_tx_cb_t cb;
 
-		__ASSERT_NO_MSG(b == pdu);
+	__ASSERT(sdu, "Invalid sdu buffer on chan %p", br_chan);
 
-		LOG_DBG("chan %p done", br_chan);
-		lower_data_ready(br_chan);
+	/* The SDU is done */
+	LOG_DBG("SDU is done, removing %p", sdu);
 
-		/* Append channel to list if it still has data */
-		if (chan_has_data(br_chan)) {
-			LOG_DBG("chan %p ready", br_chan);
-			raise_data_ready(br_chan);
-		}
+	if (!sys_slist_find_and_remove(&br_chan->_pdu_tx_queue, &sdu->node)) {
+		LOG_WRN("SDU %p is not found", sdu);
 	}
 
-	*length = pdu->len;
+	cb = closure_cb(sdu->user_data);
+	if (cb) {
+		cb(br_chan->chan.conn, closure_data(sdu->user_data), err);
+	}
 
-	return pdu;
+	/* Remove the pdu */
+	net_buf_unref(sdu);
+
+	LOG_DBG("chan %p done", br_chan);
+
+	/* Append channel to list if it still has data */
+	if (chan_has_data(br_chan)) {
+		LOG_DBG("chan %p ready", br_chan);
+		raise_data_ready(br_chan);
+	}
 }
 
-static void l2cap_br_get_info(struct bt_l2cap_br *l2cap, uint16_t info_type)
+static int bt_l2cap_br_update_req_seq_direct(struct bt_l2cap_br_chan *br_chan, uint16_t req_seq,
+					     bool rej)
 {
-	struct bt_l2cap_info_req *info;
-	struct net_buf *buf;
-	struct bt_l2cap_sig_hdr *hdr;
-
-	LOG_DBG("info type %u", info_type);
+	struct bt_l2cap_br_window *tx_win;
+	struct net_buf *sdu;
+
+	tx_win = (void *)sys_slist_peek_head(&br_chan->_pdu_outstanding);
+	while (tx_win) {
+		if (tx_win->tx_seq != req_seq) {
```

Comment:
Wouldn't the loop be simpler (or at least have less indentation), if this was
```c
while (tx_win && tx_win->tx_seq != req_seq) {
```

Reply:
Updated.